configs/configschema/coerce_value.go

```diff
@@ -225,7 +225,29 @@ func (b *Block) coerceValue(in cty.Value, path cty.Path) (cty.Value, error) {
 						elems[key.AsString()] = val
 					}
 				}
-				attrs[typeName] = cty.MapVal(elems)
+
+				// If the attribute values here contain any DynamicPseudoTypes,
+				// the concrete type must be an object.
+				useObject := false
+				switch {
+				case coll.Type().IsObjectType():
+					useObject = true
+				default:
+					// It's possible that we were given a map, and need to coerce it to an object
```

Comment:
I think the converse is possible here too... particularly if the object we're coercing is coming from decoding an HCL expression, a literal `{ ... }` is always an object, and we rely on decode-time (or coerce-time) conversion to map.

I think this can be addressed by using `ty.AttributeType(typeName)` instead of `coll.Type()` above, so it's the schema that decides which path we take, rather than the given value. (The attribute type will be `cty.DynamicPseudoType` if we need to use an object, because the schema can't predict the exact object type that will result.)